### DISCUSSION TITLE
Custom variant inside a named group with prefix

### DISCUSSION THREAD

**Discussion by martinsik** [tailwindlabs/tailwindcss#12408] [Help]
I have a repeating pattern of selectors in a group (`tw-` is my prefix):

```
<div class="tw-group/form">
  <div class="group-[.invalid:not(.untouched):not(:focus-within):not([disabled])]/form:tw-text-red">
    text
  </div>
</div>
```

This works, however I always need to repeat selector `.invalid:not(.untouched):not(:focus-within):not([disabled])`. I think I should be able to put it into a custom variant and based on this https://tailwindcss.com/docs/plugins#parent-and-sibling-states:

```
plugins: [
  function ({ addVariant, addUtilities }) {
    addVariant('group-form-invalid', ':merge(.group).invalid:not(.untouched):not(:focus-within):not([disabled]) &');
  },
],
```

Then use it as:

```
<div class="tw-group/form">
  <div class=" tw-text-blue-800 group-form-invalid/form:tw-text-red-500">
    text
  </div>
</div>
```

This seems to generate almost correct CSS selector:

```
.tw-group.tw-invalid:not(.tw-untouched):not(:focus-within):not([disabled]) .group-form-invalid\/form\:tw-text-red-500{
  --tw-text-opacity: 1;
  color: rgb(239 68 68 / var(--tw-text-opacity))
}
```

But Tailwind prefixes classes like `.invalid` or `.untouched` that come form other logic unrelated to Tailwind and I don't want them to be prefixed.
There's also `.tw-group` at the beginning that comes from I don't know where and I have no idea at this point how to change it. It should be probably `.tw-group\/form`.

Playground link: https://play.tailwindcss.com/ylj5qZrP4J?file=config

**Comment by wongjn** [comment]
You'd want to pass your plugin function as a parameter to the `plugin` function exported from `tailwindcss/plugin`:

```diff
+const plugin = require('tailwindcss/plugin');

 …

 plugins: [
-  function ({ addVariant, addUtilities }) {
+  plugin(function ({ addVariants }) {
     addVariant('group-form-invalid', ':merge(.group).invalid:not(.untouched):not(:focus-within):not([disabled]) &');
-  },
+  }),
 ],
```

One can use the [CSS white-space-separated attribute selector](https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors#attrvalue_2) to avoid the Tailwind `prefix` being applied to the inner classes:

```diff
 plugins: [
   plugin(function ({ addVariants }) {
-    addVariant('group-form-invalid', ':merge(.group).invalid:not(.untouched):not(:focus-within):not([disabled]) &');
+    addVariant('group-form-invalid', ':merge(.group)[class~="invalid"]:not([class~="untouched"]):not(:focus-within):not([disabled]) &');
   }),
 ],
```

To support the naming nomenclature, use `matchVariants()` instead of `addVariants()`:

```diff
 plugins: [
   plugin(function ({ addVariants }) {
-    addVariant('group-form-invalid', ':merge(.group)[class~="invalid"]:not([class~="untouched"]):not(:focus-within):not([disabled]) &');
+    matchVariant(
+      'group',
+      (value, { modifier }) =>
+        modifier
+          ? `:merge(.group\\/${modifier})[class~="invalid"]:not([class~="untouched"]):not(:focus-within):not([disabled]) &`
+          : ':merge(.group)[class~="invalid"]:not([class~="untouched"]):not(:focus-within):not([disabled]) &',
+      { values: { 'form-invalid': 'form-invalid' } }
+    );
   }),
 ],
```

See [this Tailwind Play](https://play.tailwindcss.com/4lNlMPyHBq?file=config).
